### PR TITLE
US-624029 : SSL and NIST SP 800-53 and NIST SP 800-131 support for Clustering Service

### PR DESCRIPTION
**Pull request by vnihal72**
US-624029 : SSL and NIST SP 800-53 and NIST SP 800-131 support Clustering Service

### CODE REVIEW

**Review by Pushpendrak18 — APPROVED** [review]

**Review by hsomu — APPROVED** [review]

**Review by punda1 — APPROVED** [review]

**Review by Pushpendrak18 — APPROVED** [review]

**Review by punda1 — APPROVED** [review]

File: charts/pega/Makefile [hunk 1]
```diff
@@ -0,0 +1,31 @@
+default: secrets
```
Comment (Saurabh-16) [inline charts/pega/Makefile:1]: When is this file invoked? Can this file be put within HZ subcharts?

File: charts/pega/templates/_pega-deployment.tpl [hunk 2]
```diff
@@ -289,6 +295,10 @@ spec:
         - name: {{ template "pegaKerberosConfig" }}-config
           mountPath: "/opt/pega/kerberos"
 {{- end }}
+{{- if (.root.Values.hazelcast.encryption.enabled) }}
+        - name: hz-encryption-secrets
+          mountPath: "/opt/hazelcast/certs"
```
Comment (Saurabh-16) [inline charts/pega/templates/_pega-deployment.tpl:300]: I assume this path is available in HZ image.
`/opt/hazelcast/certs`

Comment (Pushpendrak18) [reply charts/pega/templates/_pega-deployment.tpl:300]: yes @Saurabh-16

File: charts/pega/Makefile (hunk 1 above)
Comment (Pushpendrak18) [reply charts/pega/Makefile:1]: This file will be invoked manually, please refer: https://github.com/pegasystems/pega-helm-charts/pull/794/files#diff-346646b74f5caaa4e6ffacf9785c6718623b85624c4ad9bd81b6b4496b4d4eb2R1282
This file will generate common certificate for clustering-service and platform, so we are putting it in pega chart

**Comment by kingakowalska1** [comment]
Hi Nihal,

Here are my doc suggestions regarding the PR:

In charts/pega/README.md:

Line 1253:

`hazelcast.encryption.enabled` | Set to `true` if SSL connection is required; otherwise leave set to `false`. Note: Set this value to `false` for Pega platform versions below 24.2; if not set the installation will fail, see [this section](#optional-enabling-encryption-of-traffic-between-pega-and-clusteringservice)

To:

`hazelcast.encryption.enabled` | Set to true if you require SSL connection in your Clustering Service. Note: Highly secure encryption mode is only available in Pega Platform '24.2 and later. Set this value to false for Pega Platform release '24.1 and earlier or the installation will fail.


Line 1282:

### (Optional) Enabling encryption of traffic between Pega and ClusteringService
Run the make file in `charts/pega/Makefile` when encryption or HighlySecureCryptoMode is enabled to generate the certificates and mount it to the Hazelcast and Pega pods before Helm Install.

Parameter   | Description                                               
---         |------------------------------------------------------------
`NAMESPACE` | Namespace where Pega and Hazelcast is going to be deployed.
`CLUSTERING_SERVICE_IMAGE` | Reference the `platform/clustering-service` Docker image that you downloaded and pushed to your Docker registry that your deployment can access. 
`ENC_KEYSTORE_PASSWORD` | Key store password.
`ENC_TRUSTSTORE_PASSWORD` | Trust store password.
`HIGHLY_SECURE_CRYPTO_MODE_ENABLED` |  Enable it for complying with NIST SP 800-53 and NIST SP 800-131.

To:

### (Optional) Enabling highly secure encryption of traffic between Pega Platform and the Clustering Service
Before Helm install, run the makefile in `charts/pega/Makefile` with the following parameters when you enable encryption or HighlySecureCryptoMode to generate the certificates and mount them to the Clustering Service and Pega Platform pods.

Parameter   | Description                                               
---         |------------------------------------------------------------
`NAMESPACE` | Namespace where you deploy Pega Platform and the Clustering Service.
`CLUSTERING_SERVICE_IMAGE` | Reference the `platform/clustering-service` Docker image that you downloaded and pushed to your Docker registry that your deployment can access. 
`ENC_KEYSTORE_PASSWORD` | Keystore password.
`ENC_TRUSTSTORE_PASSWORD` | Truststore password.
`HIGHLY_SECURE_CRYPTO_MODE_ENABLED` | Set to `true` to enable the highly secure encryption mode to comply with NIST SP 800-53 and NIST SP 800-131.

And across the values.yaml files:

# Setting below to true will enforce the SSL communication between Hazelcast and Pega Infinity.

To

# Set to true to enforce SSL communication between the Clustering Service and Pega Platform.


And:

# Enable it for complying with NIST SP 800-53 and NIST SP 800-131

To:

# Set to true to comply with NIST SP 800-53 and NIST SP 800-131.

Thanks!
Kinga

**Review by Saurabh-16 — APPROVED** [review]

**Comment by taz-pega-work** [comment]
You don’t need anything from me, right? These all seem like excellent suggestions 😉
Taz

From: Kinga Kowalska ***@***.***>
Sent: Monday, August 12, 2024 6:28 AM
To: pegasystems/pega-helm-charts ***@***.***>
Cc: Aciukewicz, Taz ***@***.***>; Review requested ***@***.***>
Subject: ExtMsg: Re: [pegasystems/pega-helm-charts] US-624029 : SSL and NIST SP 800-53 and NIST SP 800-131 support for Clustering Service (PR #794)

CAUTION: This email was sent from an EXTERNAL source. Think before clicking links or opening attachments.


Hi Nihal,

Here are my doc suggestions regarding the PR:

In charts/pega/README.md:

Line 1253:

hazelcast.encryption.enabled | Set to true if SSL connection is required; otherwise leave set to false. Note: Set this value to false for Pega platform versions below 24.2; if not set the installation will fail, see this section

To:

hazelcast.encryption.enabled | Set to true if you require SSL connection in your Clustering Service. Note: Highly secure encryption mode is only available in Pega Platform '24.2 and later. Set this value to false for Pega Platform release '24.1 and earlier or the installation will fail.

Line 1282:

(Optional) Enabling encryption of traffic between Pega and ClusteringService

Run the make file in charts/pega/Makefile when encryption or HighlySecureCryptoMode is enabled to generate the certificates and mount it to the Hazelcast and Pega pods before Helm Install.
Parameter
Description
NAMESPACE
Namespace where Pega and Hazelcast is going to be deployed.
CLUSTERING_SERVICE_IMAGE
Reference the platform/clustering-service Docker image that you downloaded and pushed to your Docker registry that your deployment can access.
ENC_KEYSTORE_PASSWORD
Key store password.
ENC_TRUSTSTORE_PASSWORD
Trust store password.
HIGHLY_SECURE_CRYPTO_MODE_ENABLED
Enable it for complying with NIST SP 800-53 and NIST SP 800-131.

To:

(Optional) Enabling highly secure encryption of traffic between Pega Platform and the Clustering Service

Before Helm install, run the makefile in charts/pega/Makefile with the following parameters when you enable encryption or HighlySecureCryptoMode to generate the certificates and mount them to the Clustering Service and Pega Platform pods.
Parameter
Description
NAMESPACE
Namespace where you deploy Pega Platform and the Clustering Service.
CLUSTERING_SERVICE_IMAGE
Reference the platform/clustering-service Docker image that you downloaded and pushed to your Docker registry that your deployment can access.
ENC_KEYSTORE_PASSWORD
Keystore password.
ENC_TRUSTSTORE_PASSWORD
Truststore password.
HIGHLY_SECURE_CRYPTO_MODE_ENABLED
Set to true to enable the highly secure encryption mode to comply with NIST SP 800-53 and NIST SP 800-131.

And across the values.yaml files:

Setting below to true will enforce the SSL communication between Hazelcast and Pega Infinity.

To

Set to true to enforce SSL communication between the Clustering Service and Pega Platform.

And:

Enable it for complying with NIST SP 800-53 and NIST SP 800-131

To:

Set to true to comply with NIST SP 800-53 and NIST SP 800-131.

Thanks!
Kinga

**Comment by alemax22** [comment]
This is an enhancement that we are really interested in, but one of our security team requirement is to periodically change the encryption certificates (e.g. every 30 days). Will be possible to give the certificates by means of a secret instead of running a using some certificated created during the deployment phase? The ideal scenario will be to change the certificates, without requiring a redeploy of the entire cluster

**Comment by Pushpendrak18** [comment]
> Hi Nihal,
> 
> Here are my doc suggestions regarding the PR:
> 
> In charts/pega/README.md:
> 
> Line 1253:
> 
> `hazelcast.encryption.enabled` | Set to `true` if SSL connection is required; otherwise leave set to `false`. Note: Set this value to `false` for Pega platform versions below 24.2; if not set the installation will fail, see [this section](#optional-enabling-encryption-of-traffic-between-pega-and-clusteringservice)
> 
> To:
> 
> `hazelcast.encryption.enabled` | Set to true if you require SSL connection in your Clustering Service. Note: Highly secure encryption mode is only available in Pega Platform '24.2 and later. Set this value to false for Pega Platform release '24.1 and earlier or the installation will fail.
> 
> Line 1282:
> 
> ### (Optional) Enabling encryption of traffic between Pega and ClusteringService
> Run the make file in `charts/pega/Makefile` when encryption or HighlySecureCryptoMode is enabled to generate the certificates and mount it to the Hazelcast and Pega pods before Helm Install.
> 
> Parameter	Description
> `NAMESPACE`	Namespace where Pega and Hazelcast is going to be deployed.
> `CLUSTERING_SERVICE_IMAGE`	Reference the `platform/clustering-service` Docker image that you downloaded and pushed to your Docker registry that your deployment can access.
> `ENC_KEYSTORE_PASSWORD`	Key store password.
> `ENC_TRUSTSTORE_PASSWORD`	Trust store password.
> `HIGHLY_SECURE_CRYPTO_MODE_ENABLED`	Enable it for complying with NIST SP 800-53 and NIST SP 800-131.
> To:
> 
> ### (Optional) Enabling highly secure encryption of traffic between Pega Platform and the Clustering Service
> Before Helm install, run the makefile in `charts/pega/Makefile` with the following parameters when you enable encryption or HighlySecureCryptoMode to generate the certificates and mount them to the Clustering Service and Pega Platform pods.
> 
> Parameter	Description
> `NAMESPACE`	Namespace where you deploy Pega Platform and the Clustering Service.
> `CLUSTERING_SERVICE_IMAGE`	Reference the `platform/clustering-service` Docker image that you downloaded and pushed to your Docker registry that your deployment can access.
> `ENC_KEYSTORE_PASSWORD`	Keystore password.
> `ENC_TRUSTSTORE_PASSWORD`	Truststore password.
> `HIGHLY_SECURE_CRYPTO_MODE_ENABLED`	Set to `true` to enable the highly secure encryption mode to comply with NIST SP 800-53 and NIST SP 800-131.
> And across the values.yaml files:
> 
> # Setting below to true will enforce the SSL communication between Hazelcast and Pega Infinity.
> To
> 
> # Set to true to enforce SSL communication between the Clustering Service and Pega Platform.
> And:
> 
> # Enable it for complying with NIST SP 800-53 and NIST SP 800-131
> To:
> 
> # Set to true to comply with NIST SP 800-53 and NIST SP 800-131.
> Thanks! Kinga

Hi @kingakowalska1 , We have incorporated the suggested changes, please review/approve PR.

**Comment by kingakowalska1** [comment]
Hi, 
Since I reviewed the PR back in August, the PO driving the initiative requested that we keep a single name for this encryption enhancement, so the only comment I have is to replace every mention of "Highly secure encryption mode" with "Highly Secure Crypto Mode" in the readme.
Thanks!
Kinga

**Review by kingakowalska1 — APPROVED** [review]
Looks great, thank you for making the changes!

**Review by kishorv10 — APPROVED** [review]

**Pull request merged**